apps/wallet/src/engine/runners/solana/history.ts

```diff
@@ -0,0 +1,318 @@
+import type {
+	Connection,
+	ParsedInstruction,
+	ParsedTransaction,
+	ParsedTransactionMeta,
+	ParsedTransactionWithMeta,
+	PublicKey,
+	TokenBalance,
+} from '@solana/web3.js';
+import type { TransactionHistory } from '@walless/core';
+import { Networks } from '@walless/core';
+import { solana } from '@walless/network';
+import type { TokenDocument, TransactionHistoryDocument } from '@walless/store';
+import { selectors } from '@walless/store';
+import { storage } from 'utils/storage';
+
+import { getMetadata } from './metadata';
+import type { ParsedTokenAccountWithAddress } from './types';
+
+const historyLimit = 20;
+
+export const getTransactionsHistory = async (
+	connection: Connection,
+	wallet: PublicKey,
+	accounts: ParsedTokenAccountWithAddress[],
+) => {
+	const splTokenAccount = accounts
+		.filter((account) => account.tokenAmount.decimals > 0)
+		.map((account) => account.publicKey);
+
+	const accountList = [wallet].concat(splTokenAccount);
+	const signatureList = [
+		...new Set(await getTransactionSignatures(connection, accountList)),
+	];
+	const transactions = (
+		await getTransactionsDetail(connection, signatureList)
+	).slice(0, 20);
+	const transactionHistoryDocuments: TransactionHistoryDocument[] = (
+		await Promise.all(
+			transactions
+				.filter((transaction) => transaction !== null)
+				.map((transaction) =>
+					constructTransactionHistoryDocument(
+						connection,
+						transaction as ParsedTransactionWithMeta,
+						wallet,
+					),
+				),
+		)
+	).filter((doc) => doc !== null);
+
+	removeOldHistories(transactionHistoryDocuments);
+
+	transactionHistoryDocuments.forEach((historyDoc) => {
+		storage.upsert<TransactionHistoryDocument>(historyDoc._id, async () => {
+			return historyDoc;
+		});
+	});
+};
+
+const getTransactionSignatures = async (
+	connection: Connection,
+	accountList: PublicKey[],
+) => {
+	const signatureBufferList = await Promise.all(
+		accountList.map(async (account) => {
+			const transactionList = await connection.getSignaturesForAddress(
+				account,
+				{
+					limit: historyLimit,
+				},
+			);
+
+			return transactionList.map((trans) => trans.signature);
+		}),
+	);
+
+	const signatureList = signatureBufferList.reduce((signatureList, subList) => {
+		return signatureList.concat(subList);
+	}, []);
+
+	return signatureList;
+};
+
+const getTransactionsDetail = async (
+	connection: Connection,
+	signatureList: string[],
+) => {
+	const parsedTransactions = await connection.getParsedTransactions(
+		signatureList,
+		{
+			maxSupportedTransactionVersion: 0,
+			commitment: 'finalized',
+		},
+	);
+	const sortedTransactions = parsedTransactions.sort((trans1, trans2) =>
+		(trans1?.blockTime || 0) < (trans2?.blockTime || 0) ? 1 : -1,
+	);
+
+	return sortedTransactions;
+};
```

Comment:
I saw this function does two things, get parsed txs and sort them by time, I don't think it suits the name, `getTransactionsDetailsInBlockTimeOrder` might be suitable, or you might not need this function, just put it into the calling one because it's not reusable in another place